### PR TITLE
TRUNK-5345: Refactor ConditionService to use new functionality (Work In Progress)

### PR DESCRIPTION
**Pull request by segunolalive**
## JIRA TICKET NAME:
[TRUNK-5345: Refactor ConditionService to use new functionality](https://issues.openmrs.org/browse/TRUNK-5345)

### SUMMARY:
Refactor emrapi module's ConditionService to use this new functionality under the hood (e.g. either delegating to openmrs-core, or using the current emrapi approach, depending on the openmrs-core version)

### CODE REVIEW

**Comment by dkayiwa** [comment]
@segunolalive the new sub module should be named "api-2.2"

**Comment by segunolalive** [comment]
Okay.

**Comment by segunolalive** [comment]
@dkayiwa. Beside the package name, can I get more direction so the next update can actually be meaningful

**Comment by dkayiwa** [comment]
Just create the sub module, make sure everything compiles, and then commit.

**Comment by segunolalive** [comment]
@dkayiwa It's been updated

**Comment by segunolalive** [comment]
Done

**Comment by dkayiwa** [comment]
@segunolalive the first changes that i would expect to see for the sub module are only pom file changes.

1) Add it here: https://github.com/openmrs/openmrs-module-emrapi/blob/master/pom.xml#L36

2) There here https://github.com/openmrs/openmrs-module-emrapi/blob/master/omod/pom.xml#L46

3) And finally, an equivalent of this: https://github.com/openmrs/openmrs-module-emrapi/blob/master/api-1.12/pom.xml

Then compile to ensure that all is well. No Java changes yet.

**Comment by segunolalive** [comment]
Okay. Thanks. I'd get to that now

File: api-2.2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        <version>1.24.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>emrapi-api-2.2</artifactId>
+    <packaging>jar</packaging>
+    <name>EMR API Module API 2.2</name>
+    <description>API project for EMRAPI</description>
+
+    <properties>
+        <openMRSVersion>1.12.0</openMRSVersion>
```
Comment (dkayiwa) [inline api-2.2/pom.xml]: The openmrs version for this should be 2.2.0-SNAPSHOT

File: api-2.2/src/main/java/org/openmrs/emrapi/module/conditionslist/impl/ConditionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ *
+ * Software distributed under the License is distributed on an "AS IS"
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/emrapi/module/conditionslist/impl/ConditionServiceImpl.java]: We agreed not to include any java classes for now

Comment (segunolalive) [reply api-2.2/src/main/java/org/openmrs/emrapi/module/conditionslist/impl/ConditionServiceImpl.java]: Oh yes.I did that just be sure te imports were working correctly. Thanks. 

I'd make necessary updates immediately

File: api-2.2/pom.xml (hunk 1 above)
Comment (segunolalive) [reply api-2.2/pom.xml]: Okay.

Comment (segunolalive) [reply api-2.2/pom.xml]: Should the condition-list submodule be added as a dependency as well? I believe that's where the conditionService is implemented

Comment (dkayiwa) [reply api-2.2/pom.xml]: The condition service is coming from the platform changes that you made.

Comment (segunolalive) [reply api-2.2/pom.xml]: Okay. Thanks

**Comment by segunolalive** [comment]
@dkayiwa. It's been updated

**Comment by dkayiwa** [comment]
Have you looked at the travis failure?

**Comment by segunolalive** [comment]
Yes, It's been fixed

**Comment by dkayiwa** [comment]
Can you also add the appropriate changes here? https://github.com/openmrs/openmrs-module-emrapi/blob/master/omod/src/main/resources/config.xml#L32-L48

**Comment by segunolalive** [comment]
Done

File: omod/src/main/resources/config.xml [hunk 3]
```diff
@@ -46,6 +46,10 @@
             <path>/lib/emrapi-api-1.9-${project.parent.version}.jar</path>
             <openmrsVersion>1.9.*</openmrsVersion>
         </conditionalResource>
```
Comment (dkayiwa) [inline omod/src/main/resources/config.xml]: Can you also make the condition-list resource be for openmrs 1.9 up to 2.1?

Comment (segunolalive) [reply omod/src/main/resources/config.xml]: Okay. Done

File: omod/src/main/resources/config.xml [hunk 4]
```diff
@@ -44,7 +44,11 @@
...
         <conditionalResource>
             <path>/lib/emrapi-api-1.9-${project.parent.version}.jar</path>
-            <openmrsVersion>1.9.*</openmrsVersion>
+            <openmrsVersion>1.9.*-2.1.*</openmrsVersion>
```
Comment (dkayiwa) [inline omod/src/main/resources/config.xml]: I meant a new resource for the condition-list sub module

Comment (segunolalive) [reply omod/src/main/resources/config.xml]: Done

File: omod/src/main/resources/config.xml [hunk 5]
```diff
@@ -46,6 +46,15 @@
             <path>/lib/emrapi-api-1.9-${project.parent.version}.jar</path>
             <openmrsVersion>1.9.*</openmrsVersion>
         </conditionalResource>
+        <conditionalResource>
+            <path>/lib/emrapi-api-2.2-${project.parent.version}.jar</path>
+            <openmrsVersion>2.2.*</openmrsVersion>
+        </conditionalResource>
+        
+        <conditionalResource>
+            <path>/lib/emrapi-condition-list-${project.parent.version}.jar</path>
```
Comment (dkayiwa) [inline omod/src/main/resources/config.xml]: Can you do the same formatting as the resources before? I mean in terms of spaces and lines.

Comment (segunolalive) [reply omod/src/main/resources/config.xml]: okay. Done

File: omod/src/main/resources/config.xml [hunk 6]
```diff
@@ -46,6 +46,14 @@
             <path>/lib/emrapi-api-1.9-${project.parent.version}.jar</path>
             <openmrsVersion>1.9.*</openmrsVersion>
         </conditionalResource>
+        <conditionalResource>
+            <path>/lib/emrapi-api-2.2-${project.parent.version}.jar</path>
+            <openmrsVersion>2.2.*</openmrsVersion>
+        </conditionalResource>
+        <conditionalResource>
+            <path>/lib/emrapi-condition-list-${project.parent.version}.jar</path>
+            <openmrsVersion>1.9.*-2.1.*</openmrsVersion>
```
Comment (dkayiwa) [inline omod/src/main/resources/config.xml]: Can you make the spaces between versions be like the rest?

Comment (segunolalive) [reply omod/src/main/resources/config.xml]: Done

**Comment by dkayiwa** [comment]
Can you now compile this module with your changes and update the one you have installed? Then tell me if all is well.

**Comment by segunolalive** [comment]
Done. All is well

**Comment by dkayiwa** [comment]
Which openmrs platform version are you running?

**Comment by segunolalive** [comment]
the standalone-2.6.1

**Comment by dkayiwa** [comment]
Can you replace its war file with the latest platform version?

**Comment by segunolalive** [comment]
Okay. version 2.7, right?

**Comment by dkayiwa** [comment]
The one you worked on for the pull request that i merged.
Do some googling along these lines "difference between openmrs platform and reference application"

**Comment by segunolalive** [comment]
I'm not sure I know which PR you're referring to as I've previously worked on a number of modules 😎 . However, the google tip was pretty helpful. Just to clarify, should I run the latest stable version of the Openmrs platform or compile the openmrs-core module?

**Comment by dkayiwa** [comment]
Compile a war file from the latest snapshot version of openmrs-core

**Comment by segunolalive** [comment]
I have compiled the war file and it's running fine in the browser but I'm having trouble accessing the modules page so I can upload the omod

**Comment by segunolalive** [comment]
The omod now loads and starts successfully on the openmrs platform. What next?

**Comment by dkayiwa** [comment]
Are all your changes committed?

**Comment by segunolalive** [comment]
Yes, they are.

I just noticed the CI failure. I'm running `mvn clean install` again but it's taking forever. I think my network might be slow

**Comment by segunolalive** [comment]
The builds are passing locally

File: api-2.2/pom.xml [hunk 7]
```diff
@@ -0,0 +1,80 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.openmrs.module</groupId>
+        <artifactId>emrapi</artifactId>
+        <version>1.24.0-SNAPSHOT</version>
```
Comment (dkayiwa) [inline api-2.2/pom.xml]: Can you change this version to this before i merge it? https://github.com/openmrs/openmrs-module-emrapi/blob/master/pom.xml#L5

Comment (segunolalive) [reply api-2.2/pom.xml]: That should be 2.2-SNAPSHOT, right?

Comment (dkayiwa) [reply api-2.2/pom.xml]: Did you check the link i gave? 😊

Comment (segunolalive) [reply api-2.2/pom.xml]: Sorry, I misread your message. It's been updated now

**Comment by dkayiwa** [comment]
@segunolalive i have put a comment on JIRA for the next step

**Pull request merged**